physicsnemo/optim/combined_optimizer.py

```diff
@@ -0,0 +1,284 @@
+
+    >>> import torch
+    >>> import torch.nn as nn
+    >>> from torch.optim import Adam, SGD
+    >>> from physicsnemo.optim import CombinedOptimizer
+    >>>
+    >>> model = nn.Sequential(
+    ...     nn.Linear(10, 20),  # backbone
+    ...     nn.ReLU(),
+    ...     nn.Linear(20, 2),   # head
+    ... )
+    >>> backbone_params = list(model[0].parameters())
+    >>> head_params = list(model[2].parameters())
+    >>>
+    >>> opt1 = Adam(backbone_params, lr=1e-4)
+    >>> opt2 = SGD(head_params, lr=1e-2, momentum=0.9)
+    >>> combined_opt = CombinedOptimizer([opt1, opt2])
+    >>>
+    >>> # Use with a learning rate scheduler
+    >>> scheduler = torch.optim.lr_scheduler.StepLR(combined_opt, step_size=10)
+    >>>
+    >>> # Standard training loop
+    >>> for epoch in range(100):
+    ...     combined_opt.zero_grad()
+    ...     loss = model(torch.randn(32, 10)).sum()
+    ...     loss.backward()
+    ...     combined_opt.step()
+    ...     scheduler.step()
+    """
+
+    def __init__(
+        self,
+        optimizers: Sequence[Optimizer],
+        torch_compile_kwargs: dict[str, Any] | None = None,
+    ):
+        if not optimizers:
+            raise ValueError("`optimizers` must contain at least one optimizer.")
+
+        self.optimizers = optimizers
+        self._torch_compile_kwargs = torch_compile_kwargs
+
+        ### Aggregate parameter groups from all optimizers
+        # We pass an empty defaults dict because hyperparameters are managed by
+        # the individual optimizers, not this wrapper.
+        param_groups = [g for opt in optimizers for g in opt.param_groups]
+
+        # Flag to allow add_param_group during initialization
+        self._initializing = True
+        try:
+            super().__init__(param_groups, defaults={})
+        finally:
+            self._initializing = False
+
+        ### Setup step functions (optionally compiled)
+        if torch_compile_kwargs is None:
+            self.step_fns: list[Callable] = [opt.step for opt in optimizers]
+        else:
+            self.step_fns: list[Callable] = [
+                torch.compile(opt.step, **torch_compile_kwargs) for opt in optimizers
+            ]
+
+    def zero_grad(self, *args, **kwargs) -> None:
+        r"""Clear the gradients of all optimized parameters.
+
+        This method delegates to the ``zero_grad()`` method of each underlying
+        optimizer, passing through all arguments and keyword arguments.
+
+        Parameters
+        ----------
+        *args
+            Positional arguments to pass to each optimizer's ``zero_grad()``.
+        **kwargs
+            Keyword arguments to pass to each optimizer's ``zero_grad()``.
+            Common kwargs include ``set_to_none`` (bool).
+        """
+        for opt in self.optimizers:
+            opt.zero_grad(*args, **kwargs)
+
+    def step(
+        self, closure: Callable[[], torch.Tensor] | None = None
+    ) -> torch.Tensor | None:
```

Comment:
Upstream returns float | None, not tensor, FYI, in both the closure and `step`:

https://docs.pytorch.org/docs/stable/generated/torch.optim.Optimizer.step.html#torch.optim.Optimizer.step